datafusion/src/physical_plan/file_format/json.rs

```diff
@@ -50,6 +54,43 @@ impl NdJsonExec {
     }
 }
 
+// TODO: implement iterator in upstream json::Reader type
+struct JsonBatchReader<R: Read> {
+    reader: R,
+    schema: SchemaRef,
+    batch_size: usize,
+    proj: Option<Vec<String>>,
+}
+
+impl<R: BufRead> Iterator for JsonBatchReader<R> {
+    type Item = ArrowResult<RecordBatch>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // json::read::read_rows iterates on the empty vec and reads at most n rows
+        let mut rows = vec![String::default(); self.batch_size];
```

Comment:
This can be moved to `JsonBatchReader` and re-used across batches (that is the idea, at least)